hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -193,14 +193,16 @@ object Simplify {
 
     case ArrayLen(MakeArray(args, _)) => I32(args.length)
 
-    case ArrayLen(ToArray(StreamMap(s, _, _))) => ArrayLen(ToArray(s))
+    case ArrayLen(ToArray(StreamMap(s, _, _))) => StreamLen(s)
```

Comment:
This can work for general streams: `ArrayLen(ToArray(s)) => StreamLen(s)`. Then the rule below would handle the `StreamMap` case. And other rules like `ArrayLen(ToArray(MakeStream(...))` can be removed.

Oh, I see you have the general rule below, so this and any others that start `ArrayLen(ToArray(...))` can just be deleted.